// @gluestack-style/react/lib/module/style-sheet/index.js
import { StyledValueToCSSObject } from '../resolver';
import { getCSSIdAndRuleset } from '../updateCSSStyleInOrderedResolved.web';
import { deepMerge, resolveTokensFromConfig, addThemeConditionInMeta } from '../utils';
import { inject } from '../utils/css-injector';
export class StyleInjector {
  #globalStyleMap;
  #toBeInjectedIdsArray;
  #idCounter;
  constructor() {
    this.#globalStyleMap = new Map();
    this.#toBeInjectedIdsArray = [];
    this.#idCounter = 0;
  }
  declare(orderedSXResolved, _wrapperElementId, _styleTagId = 'css-injected-boot-time', extendedConfig) {
    const styleIds = [];
    orderedSXResolved.forEach(styledResolved => {
      var _styledResolved$meta;
      if (styledResolved !== null && styledResolved !== void 0 && (_styledResolved$meta = styledResolved.meta) !== null && _styledResolved$meta !== void 0 && _styledResolved$meta.cssId) {
        this.#globalStyleMap.set(styledResolved.meta.cssId, {
          ...styledResolved,
          type: _wrapperElementId,
          componentHash: _styleTagId,
          id: this.#idCounter,
          extendedConfig
        });
        this.#idCounter++;
        styleIds.push(styledResolved.meta.cssId);
      }
    });
    return styleIds;
  }
  resolve(cssIds = [], CONFIG, ExtendedConfig, resolve = true, declarationType = 'boot', ignoreKeys = new Set()) {
    let componentExtendedConfig = CONFIG;
    if (ExtendedConfig) {
      componentExtendedConfig = deepMerge(CONFIG, ExtendedConfig);
    }
    const toBeInjected = {};
    cssIds === null || cssIds === void 0 || cssIds.forEach(cssId => {
      if (this.#globalStyleMap.get(cssId)) {
        var _styledResolved$meta2;
        const styledResolved = this.#globalStyleMap.get(cssId);
        const theme = styledResolved === null || styledResolved === void 0 ? void 0 : styledResolved.original;
        if (resolve) {
          this.resolveComponentTheme(styledResolved, theme, componentExtendedConfig, styledResolved.componentHash, CONFIG, declarationType, ignoreKeys);
        }
        const type = styledResolved === null || styledResolved === void 0 ? void 0 : styledResolved.type;
        const styleTag = styledResolved === null || styledResolved === void 0 ? void 0 : styledResolved.componentHash;
        const cssRuleset = styledResolved === null || styledResolved === void 0 || (_styledResolved$meta2 = styledResolved.meta) === null || _styledResolved$meta2 === void 0 ? void 0 : _styledResolved$meta2.cssRuleset;
        if (!toBeInjected[type]) {
          toBeInjected[type] = new Map();
        }
        const cummialtiveCssRuleset = toBeInjected[type].get(styleTag);
        if (!cummialtiveCssRuleset) {
          toBeInjected[type].set(styleTag, {
            id: styledResolved.id,
            cssRuleset: cssRuleset ?? ''
          });
        } else {
          toBeInjected[type].set(styleTag, {
            id: cummialtiveCssRuleset === null || cummialtiveCssRuleset === void 0 ? void 0 : cummialtiveCssRuleset.id,
            cssRuleset: (cummialtiveCssRuleset === null || cummialtiveCssRuleset === void 0 ? void 0 : cummialtiveCssRuleset.cssRuleset) + cssRuleset
          });
        }
        if (styledResolved) {
          this.#globalStyleMap.set(styledResolved.meta.cssId, {
            ...styledResolved,
            value: styledResolved === null || styledResolved === void 0 ? void 0 : styledResolved.resolved
          });
        }
      }
    });
    return toBeInjected;
  }
  update(orderResolvedStyleMap) {
    const toBeInjected = {};
    orderResolvedStyleMap.forEach(styledResolved => {
      var _styledResolved$meta3;
      this.#globalStyleMap.set(styledResolved.meta.cssId, styledResolved);
      this.#toBeInjectedIdsArray.push(styledResolved.meta.cssId);
      const type = styledResolved === null || styledResolved === void 0 ? void 0 : styledResolved.type;
      const styleTag = styledResolved === null || styledResolved === void 0 ? void 0 : styledResolved.componentHash;
      const cssRuleset = styledResolved === null || styledResolved === void 0 || (_styledResolved$meta3 = styledResolved.meta) === null || _styledResolved$meta3 === void 0 ? void 0 : _styledResolved$meta3.cssRuleset;
      if (!toBeInjected[type]) {
        toBeInjected[type] = new Map();
      }
      const cummialtiveCssRuleset = toBeInjected[type].get(styleTag);
      if (!cummialtiveCssRuleset) {
        toBeInjected[type].set(styleTag, {
          id: styledResolved.id,
          cssRuleset: cssRuleset ?? ''
        });
      } else {
        toBeInjected[type].set(styleTag, {
          id: cummialtiveCssRuleset === null || cummialtiveCssRuleset === void 0 ? void 0 : cummialtiveCssRuleset.id,
          cssRuleset: (cummialtiveCssRuleset === null || cummialtiveCssRuleset === void 0 ? void 0 : cummialtiveCssRuleset.cssRuleset) + cssRuleset
        });
      }
    });
    return toBeInjected;
  }
  inject(toBeInjected = {}, inlineStyleMap) {
    Object.keys(toBeInjected).forEach(type => {
      toBeInjected[type].forEach(({
        id,
        cssRuleset
      }, styleTag) => {
        this.injectStyles(cssRuleset, type, styleTag, inlineStyleMap, id);
      });
    });
  }
  resolveComponentTheme(componentTheme, theme, componentExtendedConfig, componentHashKey, CONFIG, declarationType = 'boot', ignoreKeys = new Set()) {
    const prefixClassName = declarationType === 'inline' ? 'gs' : '';
    componentTheme.resolved = StyledValueToCSSObject(theme, componentExtendedConfig, ignoreKeys);
    addThemeConditionInMeta(componentTheme, CONFIG);

    // delete componentTheme.meta.cssRuleset;

    if (componentTheme.meta && componentTheme.meta.queryCondition) {
      var _resolveTokensFromCon;
      // console.log(
      //   JSON.parse(JSON.stringify(CONFIG)),
      //   componentTheme.meta,
      //   componentTheme.meta.queryCondition
      // );

      const queryCondition = (_resolveTokensFromCon = resolveTokensFromConfig(CONFIG, {
        condition: componentTheme.meta.queryCondition
      })) === null || _resolveTokensFromCon === void 0 ? void 0 : _resolveTokensFromCon.condition;
      // console.log(JSON.parse(JSON.stringify(CONFIG)), queryCondition);

      componentTheme.meta.queryCondition = queryCondition;
    }
    const cssData = getCSSIdAndRuleset(componentTheme, componentHashKey, prefixClassName);
    componentTheme.meta.cssRuleset = cssData.rules.style;
  }
  getStyleMap() {
    return this.#globalStyleMap;
  }
  injectStyles(cssRuleset, _wrapperType, _styleTagId, inlineStyleMap, id) {
    if (cssRuleset) {
      inject(`@media screen {${cssRuleset}}`, _wrapperType, _styleTagId, inlineStyleMap, id);
    }
  }
}
const stylesheet = new StyleInjector();
export const GluestackStyleSheet = stylesheet;
