// @gluestack-style/react/lib/commonjs/StyledProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStyled = exports.defaultConfig = exports.StyledProvider = void 0;
var _colorMode = require("./core/colorMode");
var React = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _propertyTokenMap = require("./propertyTokenMap");
var _utils = require("./utils");
var _createGlobalStylesWeb = require("./createGlobalStylesWeb");
var _createGlobalStyles = require("./createGlobalStyles");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
let colorModeSet = false;
const defaultConfig = exports.defaultConfig = {
  config: {},
  colorMode: 'light',
  components: {}
};
const defaultContextData = defaultConfig;
const StyledContext = /*#__PURE__*/React.createContext(defaultContextData);
const setCurrentColorMode = inputColorMode => {
  if (inputColorMode) {
    // console.log(get(), '>>>>>>');
    const currentColorMode = (0, _colorMode.get)();
    if (currentColorMode !== inputColorMode) {
      (0, _colorMode.set)(inputColorMode);
    }
    colorModeSet = true;
  }

  // if (inputColorMode) {
  //   set(inputColorMode === 'dark' ? 'dark' : 'light');
  //   colorModeSet = true;
  // }
};
const StyledProvider = ({
  config,
  colorMode,
  children,
  globalStyles,
  _experimentalNestedProvider
}) => {
  const inlineStyleMap = React.useRef({
    initialStyleInjected: false
  });
  inlineStyleMap.current.initialStyleInjected = false;
  const currentConfig = React.useMemo(() => {
    const configWithPlatformSpecificUnits = (0, _utils.platformSpecificSpaceUnits)(config, _reactNative.Platform.OS);
    return configWithPlatformSpecificUnits;
  }, [config]);
  if (_reactNative.Platform.OS === 'web' && globalStyles) {
    const globalStyleInjector = (0, _createGlobalStylesWeb.createGlobalStylesWeb)(globalStyles);
    globalStyleInjector({
      ...currentConfig,
      propertyTokenMap: _propertyTokenMap.propertyTokenMap
    });
  }
  const currentColorMode = React.useMemo(() => {
    return colorMode ?? (0, _colorMode.get)() ?? 'light';
  }, [colorMode]);
  const _experimentalNestedProviderRef = React.useRef(null);
  React.useEffect(() => {
    let documentElement = null;
    if (_reactNative.Platform.OS === 'web') {
      if (_experimentalNestedProvider) {
        // write own code for nested colorMode
        documentElement = _experimentalNestedProviderRef.current;
      } else {
        documentElement = document.documentElement;
      }
    }
    // Add gs class name
    if (_reactNative.Platform.OS === 'web') {
      documentElement.classList.add(`gs`);
      documentElement.classList.add(`gs-${currentColorMode}`);
    }
    (0, _colorMode.onChange)(currentColor => {
      // only for web
      if (_reactNative.Platform.OS === 'web' && !_experimentalNestedProvider) {
        const documentElement = document.documentElement;
        if (_reactNative.Platform.OS === 'web') {
          if (currentColor === 'dark') {
            documentElement.classList.remove(`gs-light`);
          } else {
            documentElement.classList.remove(`gs-dark`);
          }
          documentElement.classList.add(`gs-${currentColor}`);
        }
      }
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  React.useEffect(() => {
    setCurrentColorMode(currentColorMode);
  }, [currentColorMode]);
  React.useLayoutEffect(() => {
    if (_reactNative.Platform.OS === 'web') {
      const toBeInjectedStyles = {};
      if (inlineStyleMap.current.initialStyleInjected) {
        return;
      }
      Object.keys(inlineStyleMap.current).forEach(key => {
        if (key !== 'initialStyleInjected') {
          const styles = inlineStyleMap.current[key];
          if (!toBeInjectedStyles[key]) {
            toBeInjectedStyles[key] = document.createDocumentFragment();
          }
          styles.forEach(style => {
            if (!document.getElementById(style.id)) {
              toBeInjectedStyles[key].appendChild(style);
            }
          });
        }
      });
      Object.keys(toBeInjectedStyles).forEach(key => {
        let wrapperElement = document.querySelector('#' + key);
        if (wrapperElement) {
          wrapperElement.appendChild(toBeInjectedStyles[key]);
        }
        // delete inlineStyleMap.current[key];
      });
      inlineStyleMap.current.initialStyleInjected = true;
    }
  });
  // // Set colormode for the first time
  if (!colorModeSet) {
    setCurrentColorMode(currentColorMode);
  }
  const [animationDriverData, setAnimationDriverData] = React.useState();
  const globalStyleMap = (config === null || config === void 0 ? void 0 : config.globalStyle) && (0, _createGlobalStyles.createGlobalStyles)(config.globalStyle, _reactNative.Platform);
  const contextValue = React.useMemo(() => {
    const styledData = {
      config: currentConfig,
      globalStyle: globalStyleMap,
      animationDriverData,
      setAnimationDriverData,
      inlineStyleMap: inlineStyleMap.current
    };
    if (_experimentalNestedProvider) {
      //@ts-ignore
      styledData._experimentalNestedProvider = _experimentalNestedProvider;
      //@ts-ignore
      styledData.colorMode = colorMode;
    }
    return styledData;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentConfig, globalStyleMap, animationDriverData]);
  const providerComponent = /*#__PURE__*/React.createElement(StyledContext.Provider, {
    value: contextValue
  }, children);
  if (_experimentalNestedProvider) {
    return (
      /*#__PURE__*/
      // @ts-ignore
      React.createElement(_reactNative.View, {
        ref: _experimentalNestedProviderRef
      }, providerComponent)
    );
  } else {
    return /*#__PURE__*/React.createElement(React.Fragment, null, providerComponent);
  }
};
exports.StyledProvider = StyledProvider;
const useStyled = () => React.useContext(StyledContext);
exports.useStyled = useStyled;
